plugins/cache/cache_main.ml

```diff
@@ -6,246 +6,120 @@ include Self()
+
+  let reload () =
+    match !t with
+    | None -> read ()
+    | Some t ->
+      if t.mtime = mtime () then t.index
+      else force_read ()
+
+  let store ~f =
+    Index.with_lock ~f:(fun () ->
+        let index = reload () in
+        let index',data = f index in
+        write index';
         data)
-      ~finally:(fun () ->
-          Unix.(lockf lock F_ULOCK 0; close lock))
-
-  let update ~f = with_index ~f:(fun dir idx -> f dir idx,())
-
-  let run ~f = with_index ~f:(fun _dir idx -> idx, f idx)
-
-  let update_entry idx src entry = {
-    idx with
-    entries = Map.change idx.entries src (fun _ -> entry)
-  }
-
-  let with_entry src ~f =
-    with_index ~f:(fun _dir idx ->
-        match Map.find idx.entries src with
-        | None -> idx,None
-        | Some entry ->
-          let entry,res = f entry in
-          update_entry idx src entry,res)
 
-  let upgrade_index file version =
-    let () = match version with
-      | 1 ->
-        let old =
-          try Sexp.load_sexp file |> index_of_sexp
-          with exn ->
-            warning "can't load index: %s" (Exn.to_string exn);
-            empty in
-        index_to_file (cache_dir () / index_file) old;
-      | x ->
-        warning
-          "can't update index version from %d to %d" x index_version in
-    Sys.remove file
+  let load key ~f =
+    let idx = read () in
+    match Map.find idx.entries key with
+    | None -> None
+    | Some entry -> f entry
 
-  let versions =
-    List.map supported_versions ~f:(sprintf "index.%d")
+  let update ~f = store ~f:(fun idx -> f idx, ())
 
-  let find_index () =
-    let dir = cache_dir () in
-    let files = List.map versions ~f:(fun x -> dir / x) in
-    List.find files ~f:Sys.file_exists
-
-  let upgrade () =
-    match find_index () with
-    | None -> ()
-    | Some file -> match get_version file with
-      | Ok ver when Int.(ver = index_version) -> ()
-      | Ok ver -> upgrade_index file ver
-      | Error er ->
-        error "unknown index version: %s" (Error.to_string_hum er)
+  let iter ~f = match !t with
+    | None -> f (read ())
+    | Some {index} -> f index
 end
 
-let size file =
-  Unix.LargeFile.((stat file).st_size)
+let size file = Unix.LargeFile.((stat file).st_size)
 
 let cleanup () =
-  Index.update ~f:(fun _ idx ->
+  Global.update ~f:(fun idx ->
+      Map.iter idx.entries ~f:GC.remove_entry;
       {idx with entries = Data.Cache.Digest.Map.empty});
   exit 0
 
 let set_size size =
-  Index.update ~f:(fun _ idx ->
-      {idx with config = {max_size = Int64.(size * 1024L * 1024L)}});
+  Global.update ~f:(fun idx ->
+      let max_size = Int64.(size * 1024L * 1024L) in
+      {idx with config = {idx.config with max_size;}})
+
+let set_overhead overhead =
+  if Float.(overhead >= 0.0 && overhead < 1.0)
+  then
+    Global.update ~f:(fun idx ->
+        {idx with config = {idx.config with overhead}})
+  else
+    raise (Invalid_argument "Cache overhead should be in the range [0.0; 1.0) ")
```

Comment:
No, it is totally fine for the overhead to be 1000% or whatever large number. There is no need to cap it with 100%